swift/apple/FirezoneKit/Sources/FirezoneKit/Managers/TunnelManager.swift

```diff
@@ -0,0 +1,258 @@
+
+  // Use separate bundle IDs for release and debug.
+  // Helps with testing releases and dev builds on the same Mac.
+  #if DEBUG
+    private let bundleIdentifier = Bundle.main.bundleIdentifier.map {
+      "\($0).debug.network-extension"
+    }
+
+    private let bundleDescription = "Firezone (Debug)"
+  #else
+    private let bundleIdentifier = Bundle.main.bundleIdentifier.map { "\($0).network-extension" }
+    private let bundleDescription = "Firezone"
+  #endif
+
+  init() {
+    manager = nil
+  }
+
+  // Initialize and save a new VPN profile in system Preferences
+  func create() async throws -> Settings {
+    let protocolConfiguration = NETunnelProviderProtocol()
+    let manager = NETunnelProviderManager()
+    let settings = Settings.defaultValue
+
+    protocolConfiguration.providerConfiguration = settings.toProviderConfiguration()
+    protocolConfiguration.providerBundleIdentifier = bundleIdentifier
+    protocolConfiguration.serverAddress = settings.apiURL
+    manager.localizedDescription = bundleDescription
+    manager.protocolConfiguration = protocolConfiguration
+
+    // Save the new VPN profile to System Preferences and reload it,
+    // which should update our status from invalid -> disconnected.
+    // If the user denied the operation, the status will be .invalid
+    try await manager.saveToPreferences()
+    try await manager.loadFromPreferences()
+
+    await statusChangeHandler?(manager.connection.status)
+    self.manager = manager
+
+    return settings
+  }
+
+  func load(callback: @escaping (NEVPNStatus, Settings?, String?) -> Void) {
+    Task {
+      // loadAllFromPreferences() returns list of tunnel configurations created by our main app's bundle ID.
+      // Since our bundle ID can change (by us), find the one that's current and ignore the others.
+      guard let managers = try? await NETunnelProviderManager.loadAllFromPreferences()
+      else {
+        Log.app.error("\(#function): Could not load VPN configurations!")
+        return
+      }
+
+      Log.app.log("\(#function): \(managers.count) tunnel managers found")
+      for manager in managers {
+        if let protocolConfiguration = manager.protocolConfiguration as? NETunnelProviderProtocol,
+           protocolConfiguration.providerBundleIdentifier == bundleIdentifier,
+           let providerConfiguration = protocolConfiguration.providerConfiguration as? [String: String]
+        {
+          // Found it
+          let settings = Settings.fromProviderConfiguration(providerConfiguration)
+          let actorName = providerConfiguration[TunnelManagerKeys.actorName]
+          let status = manager.connection.status
+
+          // Share what we found with our caller
+          callback(status, settings, actorName)
+
+          // Update our state
+          self.manager = manager
+
+          // Stop looking for our tunnel
+          break
+        }
+      }
+
+      // Hook up status updates
+      setupTunnelObservers()
```

Comment:
This is harmless if we can't find our tunnel?

Reply:
Yeah this just sets up a bridge to receive notifications from our tunnel and act upon them -- we should either find our tunnel (which corresponds to our bundle ID) or create it